### PR TITLE
Add ephemeral key allocator to FabricTable

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem
- CASE requires P256 ephemeral keys
- CASE had a hack whereby "single slots" ephemeral keys
  for NXP HSM were used, which cannot work for multiple CASE
  session establishments
- Using raw P256Keypair prevents being able to use OS-backed
  or HW-backed keys, like can be done for operational keys

Issue #20036

#### Change overview
- Adds a way to get CASE ephemeral keys from the OperationalKeystore
  abstraction
- Funnels their access via the FabricTable
- Removes some HSM hacks (cannot remove all HSM usage just yet)
  in a way that now OperationalKeystore targeting NXP HSM could
  do the right thing

#### Testing

- Unit tests still pass
- Integration tests still passa
- Added unit tests to validate usage of new APIs
- Ran valgrind on the unit tests, found no leaks

### CODE REVIEW

File: src/crypto/OperationalKeystore.h [hunk 1]
```diff
@@ -168,6 +168,26 @@ class OperationalKeystore
...
+     * @brief Create an ephemeral keypair for use in session establishment.
+     *
+     * WARNING: There may be more than one of these in use concurrently, so implementations have to
+     *          be careful to avoid assuming a single instance.
+     *
+     * WARNING: The return value MUST be released by `ReleaseEphemeralKeypair`. This is because
+     *          Matter CHIPMem.h does not properly support UniquePtr in a way that would
+     *          safely allow classes derived from Crypto::P256Keypair to be released properly.
+     *
+     * @return a pointer to a dynamically P256Keypair (or derived class thereof), which may evaluate to nullptr
```
Comment (bzbarsky-apple) [inline src/crypto/OperationalKeystore.h:182]:
```suggestion
     * @return a pointer to a P256Keypair (or derived class thereof), which may evaluate to nullptr
```
Nothing prevents the keystore from allocating these however it wants (e.g. a static fixed-size array).

File: src/crypto/OperationalKeystore.h [hunk 2]
```diff
@@ -168,6 +168,26 @@ class OperationalKeystore
...
+     *         if running out of memory.
+     */
+    virtual Crypto::P256Keypair * AllocateEphemeralKeypair() = 0;
```
Comment (bzbarsky-apple) [inline src/crypto/OperationalKeystore.h:185]: This should document that the caller is expected to call `Initialize()`, right?

That said, why is the caller expected to do that?  Can't the implementation here handle any needed initialization?

File: src/protocols/secure_channel/CASESession.h [hunk 3]
```diff
@@ -26,16 +26,14 @@
...
-#include <crypto/hsm/CHIPCryptoPALHsm.h>
-#endif
 #include <credentials/CertificateValidityPolicy.h>
 #include <credentials/FabricTable.h>
 #include <credentials/GroupDataProvider.h>
+#include <crypto/CHIPCryptoPAL.h>
 #include <lib/core/CHIPTLV.h>
 #include <lib/core/ScopedNodeId.h>
 #include <lib/support/Base64.h>
+#include <lib/support/CHIPMem.h>
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASESession.h:36]: Why do we need CHIPMem in the header here?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/crypto/OperationalKeystore.h (hunk 2 above)
Comment (tcarmelveilleux) [reply src/crypto/OperationalKeystore.h:185]: Initialize is not needed when deserializing. I kept the existing convention of not assuming Initialize, which is the caller's role to do. Initialize costs an EC key pair generation, which doesn't have zero cost.

File: src/protocols/secure_channel/CASESession.h (hunk 3 above)
Comment (tcarmelveilleux) [reply src/protocols/secure_channel/CASESession.h:36]: At first I had a UniquePtr until it was proven that I can't use that for the purpose. Forgot to remove the header. Done now.

File: src/crypto/OperationalKeystore.h (hunk 1 above)
Comment (tcarmelveilleux) [reply src/crypto/OperationalKeystore.h:182]: Fixed the wording.

File: src/crypto/OperationalKeystore.h (hunk 2 above)
Comment (bzbarsky-apple) [reply src/crypto/OperationalKeystore.h:185]: But are there any cases where someone does `AllocateEphemeralKeypair` and then deserializes into it, instead of doing an `Initialize`?

Comment (tcarmelveilleux) [reply src/crypto/OperationalKeystore.h:185]: No, I don't think so. The only usage is for one to be initialized. I could add that change as a follow-up, or I will do it if CI fails and I need to repush

**Review by andy31415 — APPROVED** [review]

File: src/crypto/OperationalKeystore.h (hunk 2 above)
Comment (tcarmelveilleux) [reply src/crypto/OperationalKeystore.h:187]: Done.

File: src/crypto/OperationalKeystore.h [hunk 4]
```diff
@@ -168,6 +168,28 @@ class OperationalKeystore
...
     virtual CHIP_ERROR SignWithOpKeypair(FabricIndex fabricIndex, const ByteSpan & message,
                                          Crypto::P256ECDSASignature & outSignature) const = 0;
+
+    /**
+     * @brief Create an ephemeral keypair for use in session establishment.
+     *
+     * The caller must Initialize() the P256Keypair if needed. It is not done by this method.
+     *
+     * NOTE: There may be more than one of these in use concurrently, so implementations have to
+     *       be careful to avoid assuming a single instance.
```
Comment (msandstedt) [inline src/crypto/OperationalKeystore.h:178]: Is this stating that implementations must support concurrent allocation of more than one keypair, or that implementations must ensure that no more keypairs are concurrently allocated than the underlying allocator can support?

If the former, why not just state more obviously: the stack will try to allocate more than one, so you better support this.

If the latter, this seems to lack guidance on how implementations would actually achieve enforcing that no more key pairs then they can support will be allocated.

Comment (tcarmelveilleux) [reply src/crypto/OperationalKeystore.h:178]: The former. Let me rephrase

**Review by msandstedt — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

File: src/protocols/secure_channel/CASESession.h [hunk 5]
```diff
@@ -269,11 +262,8 @@ class DLL_EXPORT CASESession : public Messaging::UnsolicitedMessageHandler,
 
     Crypto::Hash_SHA256_stream mCommissioningHash;
     Crypto::P256PublicKey mRemotePubKey;
-#ifdef ENABLE_HSM_CASE_EPHEMERAL_KEY
-    Crypto::P256KeypairHSM mEphemeralKey;
-#else
-    Crypto::P256Keypair mEphemeralKey;
-#endif
+    Crypto::P256Keypair * mEphemeralKey = nullptr;
+    Crypto::P256Keypair mEphemeralKey2;
```
Comment (tehampson) [inline src/protocols/secure_channel/CASESession.h:266]: What is this used for?

**Review by tehampson — APPROVED** [review]

File: src/protocols/secure_channel/CASESession.h (hunk 5 above)
Comment (tcarmelveilleux) [reply src/protocols/secure_channel/CASESession.h:266]: Good catch, that was a side effect of testing a test
